glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+    ///Requested access can not be granted till already hold access will be released
+    WouldBlock,
+}
+
+impl Display for TryReadWriteLockError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        match self {
+            TryReadWriteLockError::Closed(error) => Display::fmt(error, f),
+            TryReadWriteLockError::WouldBlock => f.write_str("Call would block execution"),
+        }
+    }
+}
+
+impl Debug for TryReadWriteLockError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        match self {
+            TryReadWriteLockError::Closed(error) => Debug::fmt(error, f),
+            TryReadWriteLockError::WouldBlock => f.write_str("WouldBlock"),
+        }
+    }
+}
+
+impl Error for TryReadWriteLockError {}
+
+impl From<LockClosedError> for TryReadWriteLockError {
+    fn from(error: LockClosedError) -> Self {
+        TryReadWriteLockError::Closed(error)
+    }
+}
+
+struct Waiter<T> {
+    kind: WaiterKind,
+    id: WaiterId,
+    rw: Rc<RefCell<State<T>>>,
+}
+
+/// A reader-writer lock
+///
+/// This type of lock allows a number of readers or at most one writer at any
+/// point in time. The write portion of this lock typically allows modification
+/// of the underlying data (exclusive access) and the read portion of this lock
+/// typically allows for read-only access (shared access).
+///
+/// An `ReadWriteLock` will allow any number of readers to acquire the
+/// lock as long as a writer is not holding the lock.
+///
+/// The priority policy of the lock is based on FIFO policy. Fibers will be granted access in the
+/// order in which access to the lock was requested.
+///
+/// Lock is not reentrant, yet. That means that two subsequent calls to request write access
+/// to the lock will lead to dead lock problem.
+///
+/// The type parameter `T` represents the data that this lock protects. The RAII guards
+/// returned from the locking methods implement [`Deref`] (and [`DerefMut`]
+/// for the `write` methods) to allow access to the content of the lock.
+///
+///
+/// # Examples
+///
+/// ```
+///use glommio::read_write_lock::ReadWriteLock;
+///use glommio::LocalExecutor;
+///
+/// let lock = ReadWriteLock::new(5);
+/// let ex = LocalExecutor::make_default();
+///
+/// ex.run( async move {
+///    // many reader locks can be held at once
+///    {
+///        let r1 = lock.read().await.unwrap();
+///        let r2 = lock.read().await.unwrap();
+///        assert_eq!(*r1, 5);
+///        assert_eq!(*r2, 5);
+///    } // read locks are dropped at this point
+///
+///    // only one write lock may be held, however
+///    {
+///        let mut w = lock.write().await.unwrap();
+///        *w += 1;
+///        assert_eq!(*w, 6);
+///    } // write lock is dropped here
+/// });
+/// ```
+///
+/// [`Mutex`]: super::Mutex
+#[derive(Debug)]
+pub struct ReadWriteLock<T> {
+    rw: Rc<RefCell<State<T>>>,
+}
+
+#[derive(Debug)]
+struct State<T> {
+    value: UnsafeCell<T>,
```

Comment:
Yeah, given that this *already* is behind `RefCell`, I don't think an `UnsafeCell` is needed? I think we'll need a smudge of unsafe code here to extend lifetimes past `Ref` for guards, that can be done without UnsafeCell probably. 

Cursory look, might be wrong here :-)